### ISSUE TITLE
[Proposal] segmentMetadata query returns full list of dimensions

### ISSUE THREAD

**Issue by VladimirIordanov** [apache/incubator-druid#7583]
### Motivation

Currently segmentMetadata query doesn't return dimensions which have no values (cardinality=0) for the time range queried. Some tools (i.e. Imply Pivot) use this type of query to fetch metadata and then validate all user requests against it. When these tools got in such situation with 'non-existing' dimension they treat the request as invalid. Although Druid Broker can process JSON request with such dimension correctly and return an expected result. Even more that these dimensions are displayed in timeline in coordinator views. So there is even inconsistency between it and segmentMetadata query.
Situation got worse after introducing SQL to Druid. For now Druid fails when these dimensions are requested in SQL expression. Example of the error response:
{
    "error": "Unknown exception",
    "errorMessage": "org.apache.calcite.runtime.CalciteContextException: From line 1, column 8 to line 1, column 18: Column 'xxx' not found in any table",
    "errorClass": "org.apache.calcite.tools.ValidationException",
    "host": null
}
It looks like Apache Calcite tries to follow the described workflow and match request to metadata and receives the same trouble.

### Proposed changes

Change segmentMetadata query to output all dimensions even they do not have data in the requested time range.

### Rationale

It is natural for data storages to have such dimensions with occasional data. The metadata of an data source should be the same for any time range even there are such dimensions in it. Behavior of Druid should be consistent across all its tools and do not change on any cases. Druid core should be consistent and SQL queries should not fail an error for requests which JSON equivalent works as expected.

### Operational impact

No operational impact.

**Comment by sascha-coenen** [comment]
This is an awesome feature request!

This has been an ongoing issue on our side as well.
How sparse a dimensions should not make a given query fail.

If a dimension contains a value in 1000 cases, queries work and the dimension is considered part of the schema. Same thing if the dimension contains 1 single non-null record and therefore has cardinality 1.
If this is permissible, then it must not be the case that if the cardinality drops from 1 to, which is a very small difference, that this minute difference is a deciding factor for whether a dimension makes it into the schema or not, whether an SQL query fails or succeeds.

From my user's perspective I applaud this proposal and believe that that indeed this current behaviour aught to be changed.

I'm afraid that people might be hesitant to make this change thought, reason being that changing the behaviour of the metadataSegment query might be seen as a breaking change that is not backwards compatible.
However, one might introduce a boolean flag into the request schema of a metadataSegment query that indicates whether the automatic schema construction should omit fields of cardinality=0 or include them. By setting the default to omit, one stays backwards compatible but yet allows folks to have sparse dimensions.

As native queries do NOT fail on dimensions with cardinality 0 and even work for non-existing dimensions because their semantics are also well defined (cardinality=0, min=null, max=null, count=0, etc), I would also propose to fix the situation that SQL queries fail on cardinality=0 dimensions and would even suggest that they also should work on non-existing fields.
However, whether SQL should fail or succeed on non-existing fields is highly debated, so it would be okay to let an SQL statement fail IF a field does not exist.
BUT the issue is that currently SQL statements are failing even on fields that DO exist and are even listed in the segment dimension lists available via the coordinator console as Vladimir pointed out, and yet the SQL query fail.
At the very least, this behaviour is inconsistent and I would agree to resolve this inconsistency by accepting a dimension with 0 entries as an existing dimension with cardinality 0 and count 0.

**Comment by gianm** [comment]
> Situation got worse after introducing SQL to Druid. For now Druid fails when these dimensions are requested in SQL expression. Example of the error response:

Fwiw, Druid SQL checks segment metadata for _all_ segments in a datasource (not just the time range being queried), so this 'column not found' error should only happen if a column has no values at any point, in any segment.

> Change segmentMetadata query to output all dimensions even they do not have data in the requested time range.

This wouldn't help with the Druid SQL issue, due to the reason mentioned above.

The heart of the issue is the difference between the list of dimensions that one has configured at ingestion time, and the list of columns that are actually physically stored within the Druid segments. Druid won't store columns that don't have any data, as an optimization to avoid storing a column full of nulls. So I think there are two ways to address this:

1. Change things such that Druid _does_ store columns that don't have any data, ideally in a special lightweight way so they don't take up much space (beyond some metadata that the column exists and is empty).
2. Change things such that Druid stores the list of configured dimensions in the segment `Metadata` object. (It already stores aggregators and queryGranularity in here, which is how those features of the segmentMetadata work.) Then, the segmentMetadata queries could return the list of configured dimensions, which would potentially be more comprehensive than the list of columns that actually got stored. Also, change Druid SQL to consider these dimensions-that-don't-have-physical-columns as valid SQL columns.

I'd probably recommend (2) since this dimensions list might be useful for other purposes too, and it doesn't require changes to the segment storage format, so it doesn't have any backwards-compatibility concerns.

**Comment by VladimirIordanov** [comment]
Thank you Gian for the fast response. In my understanding the second option is more safe in terms of backward compatibility. So my vote is also for it. Unfortunately I know Druid from user perspective and don't familiar with its codebase. How big efforts to implement it? Will it require lot of work? Hope not much and I could see this feature implemented in the nearest release.
Thank you very much for the support.

**Comment by sascha-coenen** [comment]
same here.

I noticed that there is a segment format V10 project and perhaps changes such as in option 1 might be picked up in that context, but to address the current issue, option 2 sounds like an easier, less invasive route.

**Comment by gianm** [comment]
> Thank you Gian for the fast response. In my understanding the second option is more safe in terms of backward compatibility. So my vote is also for it. Unfortunately I know Druid from user perspective and don't familiar with its codebase. How big efforts to implement it? Will it require lot of work? Hope not much and I could see this feature implemented in the nearest release.

I don't think it would be too big, if you are interested in contributing then the place to start would be adding the dimensionsSpec to org.apache.druid.segment.Metadata, and then plumbing it into the segmentMetadata query using a new analysisType.
